cmd/file_lint.go

```diff
@@ -0,0 +1,165 @@
+	}
+
+	var outputFormat string
+	{
+		outputFormat, err = cmd.Flags().GetString("format")
+		if err != nil {
+			return fmt.Errorf("failed getting cli argument 'format'; %w", err)
+		}
+		outputFormat = strings.ToUpper(outputFormat)
+	}
+
+	inputRuleset, err := cmd.Flags().GetString("ruleset")
+
+	if err != nil || inputRuleset == "" {
+		return fmt.Errorf("failed getting cli argument 'ruleset';")
+	}
+
+	ruleSetBytes, err := ioutil.ReadFile(inputRuleset)
+	if err != nil {
+		panic(err.Error())
+	}
+
+	customRuleSet, rsErr := rulesets.CreateRuleSetFromData(ruleSetBytes)
+
+	if rsErr != nil {
+		panic(err.Error())
+	}
+
+	specBytes, err := ioutil.ReadFile(inputFilename)
+	if err != nil {
+		panic(err.Error())
+	}
+
+	ruleSetResults := motor.ApplyRulesToRuleSet(&motor.RuleSetExecution{
+		RuleSet:           customRuleSet,
+		Spec:              specBytes,
+		SkipDocumentCheck: true,
+	})
+
+	failingCount := 0
+	totalCount := 0
+
+	var lintResults []map[string]interface{}
+	for _, x := range ruleSetResults.Results {
+
+		if onlyFailures && (x.Rule.Severity != "error") {
+			continue
+		}
+
+		if severityList[x.Rule.Severity] >= severityList[minimumSeverity] {
+			failingCount++
+		}
+
+		totalCount++
+
+		lintResults = append(lintResults, map[string]interface{}{
+			"message":  x.Message,
+			"path":     x.Rule.Given,
+			"line":     x.StartNode.Line,
+			"column":   x.StartNode.Column,
+			"severity": x.Rule.Severity,
+		})
+	}
+
+	lintErrs := map[string]interface{}{
+		"total_count": totalCount,
+		"fail_count":  failingCount,
+		"results":     lintResults,
+	}
+
+	if outputFormat == "DEFAULT" {
+		if totalCount > 0 {
+			fmt.Printf("Linting Violations: %d\n", totalCount)
+			fmt.Printf("Failures: %d\n\n", failingCount)
+			for _, violation := range lintErrs["results"].([]map[string]interface{}) {
+				fmt.Printf("[%d:%d] %s\n", violation["line"], violation["column"], violation["message"])
+			}
+		}
+	} else {
+		err = filebasics.WriteSerializedFile("-", lintErrs, filebasics.OutputFormat(outputFormat))
+		if err != nil {
+			return err
+		}
+	}
+
+	if failingCount > 0 {
+		// We don't want to print the error here as they're already output above
+		// But we _do_ want to set an exit code of failure
+		cmd.SilenceErrors = true
+		return errors.New("linting errors detected")
+	}
+
+	return nil
+}
+
+//
+//
+// Define the CLI data for the lint command
+//
+//
+
+func newLintCmd() *cobra.Command {
+	lintCmd := &cobra.Command{
+		Use:   "lint [flags] lint-file",
+		Short: "Short description here",
+		Long:  `Long description Here`,
+		RunE:  executeLint,
+	}
+
+	lintCmd.Flags().StringVarP(&cmdLintInputFilename, "state", "s", "-",
+		"decK file to process. Use - to read from stdin.")
+	lintCmd.Flags().StringVarP(&cmdLintInputRuleset, "ruleset", "r", "",
+		"Ruleset to apply to the state file.")
+	lintCmd.Flags().StringP("format", "", "default", "output format: default, "+
```

Comment:
Providing a `--format` different than `json` or `yaml` returns the following error:

```
$ ./deck file lint -s kong.yaml --ruleset ruleset.yaml --format foooo
Error: expected 'format' to be either 'yaml' or 'json', got: 'foooo'
```

The error suggests these two are the only allowed formats, while `default` is also supported.

Reply:
I don't think we should support `--format default`, that's ambiguous. "default" should be specified in the help documentation as to what happens when you don't provide a value of either `json` or `yaml` for `--format`

Reply:
I agree `--format default` shouldn't be supported, but I also think `default` is not a good name for this. If we change it to a better name, then the message may have this other option too. WDYT?

Reply:
I think this flag should be:
```
--format string          Output format: yaml or json. (default "yaml")
```

which matches the other deck file commands

Reply:
So, just to clarify, no "human readable" option?

Reply:
I think we need a human readable option. Agree with renaming away from default. The `plaintext` suggestion seems reasonable to me

Reply:
> So, just to clarify, no "human readable" option?

Sorry, I was confused by the purpose of the flag here, I understand now.   I agree @GGabriele, but maybe a value of `plain` is clear enough?  Or `plaintext' also works